### PR TITLE
Add Zphinx & further updates

### PR DESCRIPTION
**Pull request by rolfschr**
This PR was initially only to include [zphinx-zerver](https://github.com/stef/zphinx-zerver). However, things happen*. Thus, this PR now includes:

* Update pwdsphinx & androsphinx (now both with the new upstream dep "equihash")
* Add zphinx-zerver (a sphinx server impl)
* Make package definitions more consistent
* Rework flake checks

Note that [androsphinx is currently unable to communicate with the server implementations](https://github.com/dnet/androsphinx/issues/9). However, this is "their" problem and should not interfere with this PR - I expect a simple androsphinx source code revision update on the Nix side.

\*  This repo contains the code for https://github.com/ngi-nix/ngi/issues/89 (this PR) as well as https://github.com/ngi-nix/ngi/issues/46. I first worked on #46, later on #89. Work on #89 was paused though because I first wanted to incorporated (any possible major) changes from #46's PR feedback. In the meantime, the project authors continued add breaking changes. Thus, this PR became larger.

### CODE REVIEW

File: pkgs/androsphinx/default.nix [hunk 1]
```diff
@@ -1,6 +1,8 @@
-{ version, src, androsphinxCryptoLibs, sdk, callPackage }:
+{ pkgs, version, src, androsphinxCryptoLibs, sdk }:
+with pkgs;
 let buildGradle = callPackage ./gradle-env.nix { };
 in buildGradle {
+  pname = "androsphinx";
```
Comment (rolfschr) [inline pkgs/androsphinx/default.nix:5]: Changes in this file and others are made for consistency only.

File: pkgs/pwdsphinx/default.nix [hunk 2]
```diff
@@ -18,14 +20,13 @@ buildPythonPackage rec {
...
     FILE=sphinx.cfg
     cp sphinx.cfg_sample $FILE
-    for SECTION in client server ; do
-      for PREFIX in ssl_key ssl_cert ; do
-        crudini --set $FILE $SECTION $PREFIX ./''${PREFIX}.doesnt-exist.pem
-      done
+    # Some settings are commented out but must be activated.
+    # "# address=..." -> "address=..."
+    for PREFIX in address timeout ; do
+      sed -i "s/^#[[:space:]]*$PREFIX=/$PREFIX=/" $FILE
```
Comment (rolfschr) [inline pkgs/pwdsphinx/default.nix:28]: The sample config file in `src` has evolved ...

File: pkgs/zphinxzerver/default.nix [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  buildInputs = [ libsodium.dev rename zig ];
+
+  postPatch = ''
+    substituteInPlace build.zig \
+      --replace 'exe.addIncludeDir(".");' \
+      'exe.addIncludeDir("."); exe.addIncludeDir("${libsodium.dev}/include"); exe.addIncludeDir("${equihash}/include");'
+
+    substituteInPlace build.zig \
+      --replace 'exe.addLibPath("/usr/lib");' \
+      'exe.addLibPath("/usr/lib"); exe.addLibPath("${equihash}/lib");'
```
Comment (rolfschr) [inline pkgs/zphinxzerver/default.nix:16]: I'm not so happy with this patching here. It feels somewhat ugly. Here is the [original file](https://github.com/stef/zphinx-zerver/blob/master/build.zig). We need to make sure that the zig compiler finds our libsodium & equihash libs.

File: pkgs/pwdsphinx/default.nix [hunk 4]
```diff
@@ -18,14 +20,13 @@ buildPythonPackage rec {
 
   checkInputs = [ crudini ];
```
Comment (ysndr) [inline pkgs/pwdsphinx/default.nix]: I think this can be removed now that sed is used right?

File: pkgs/pysodium/default.nix [hunk 5]
```diff
@@ -1,4 +1,5 @@
-{ version, src, stdenv, buildPythonPackage, libsodium }:
+{ pkgs, version, src }:
```
Comment (ysndr) [inline pkgs/pysodium/default.nix]: With native python libs one can easily run into python version incompatibilities, using the old approach allows to use different python versions to build at least when picked from an overlay.

was there a specific reason to opt for this new format?

File: pkgs/zphinxzerver/default.nix (hunk 3 above)
Comment (ysndr) [reply pkgs/zphinxzerver/default.nix:16]: I assume what is ugly here is relying on arbitrary strings (`exe.addIncludeDir(".");`) right?
For that one can use patch files to insert anchors which would then be replaced with substituteInPlace

**Review by ysndr — COMMENTED** [review]
The nix side of this looks good so far except a few comments i've left
I'll try building the server later but don't see any reason that it would fail right now.

File: pkgs/pysodium/default.nix (hunk 5 above)
Comment (rolfschr) [reply pkgs/pysodium/default.nix]: I wanted to make sure that all packages are built using the same nixpkgs/python version. I understand you point though. I guess you would simply keep the ``buildPythonPackage`` parameter ?
(Btw: [Here](https://github.com/ngi-nix/opaque-sphinx/pull/2/files#diff-206b9ce276ab5971a2489d75eb1b12999d4bf3843b7988cbe8d687cfde61dea0R134) the function is declared and used implicitly. I guess it's actually good to make that more explicit.)

Edit: Improved link: https://github.com/ngi-nix/opaque-sphinx/blob/zphinx/flake.nix#L134

File: pkgs/pwdsphinx/default.nix (hunk 4 above)
Comment (rolfschr) [reply pkgs/pwdsphinx/default.nix]: Yes, thanks!

**Comment by rolfschr** [comment]
@ysndr Thanks for the quick review and your comments. I changed as suggested.

**Pull request merged**